File: sdk/php/tests/regression/RunForeverTest.php
<?php

require 'vendor/autoload.php';

use KuCoin\UniversalSDK\Api\DefaultClient;
use KuCoin\UniversalSDK\Generate\Spot\Market\GetAllSymbolsReq;
use KuCoin\UniversalSDK\Model\ClientOptionBuilder;
use KuCoin\UniversalSDK\Model\Constants;
use KuCoin\UniversalSDK\Model\TransportOptionBuilder;
use KuCoin\UniversalSDK\Model\WebSocketClientOptionBuilder;
use React\EventLoop\Factory as Loop;

final class Runner
{
    private const STEP = 5.0;

    private int $wsMsgCnt = 0;
    private int $wsErrCnt = 0;
    private int $mkErrCnt = 0;

    private $wsSvc;
    private $marketApi;
    private $loop;

    public function __construct()
    {
        $this->loop = Loop::create();
        $cli = new DefaultClient(
            (new ClientOptionBuilder())
                ->setKey(getenv('API_KEY') ?: '')
                ->setSecret(getenv('API_SECRET') ?: '')
                ->setPassphrase(getenv('API_PASSPHRASE') ?: '')
                ->setTransportOption(
                    (new TransportOptionBuilder())
                        ->setKeepAlive(true)
                        ->build()
                )
                ->setWebSocketClientOption((new WebSocketClientOptionBuilder())->build())
                ->setSpotEndpoint(Constants::GLOBAL_API_ENDPOINT)
                ->setFuturesEndpoint(Constants::GLOBAL_FUTURES_API_ENDPOINT)
                ->setBrokerEndpoint(Constants::GLOBAL_BROKER_API_ENDPOINT)
                ->build()
            , $this->loop
        );

        $this->wsSvc = $cli->wsService();
        $this->marketApi = $cli->restService()->getSpotService()->getMarketApi();

    }

    public function run(): void
    {
        $this->marketLoop();
        $this->wsForever();
        $this->wsStartStopLoop();
        $this->statLoop();
        $this->loop->run();
    }

    private function marketLoop(): void
    {
        $this->loop->addPeriodicTimer(self::STEP, function () {
            try {
                $resp = $this->marketApi
                    ->getAllSymbols(GetAllSymbolsReq::builder()->setMarket('USDS')->build());
                printf("MARKET API [OK] %d \n", count($resp->data));
            } catch (\Throwable $e) {
                $this->mkErrCnt++;
                echo "MARKET API [ERROR] {$e->getMessage()}\n";
            }
        });
    }

    private function wsForever(): void
    {
        $ws = $this->wsSvc->newSpotPublicWs();
        $ws->start()->then(function () use ($ws) {
            return $ws->orderbookLevel50(['ETH-USDT', 'BTC-USDT'], function ($data) {
                $this->wsMsgCnt++;
            });
        })->then(function () {
            echo "WS [OK]\n";
        })->catch(function ($err) {
            echo "WS [Error] " . $err->getMessage() . "\n";
        });

    }

    private function wsStartStopLoop(): void
    {
        $this->loop->addPeriodicTimer(self::STEP, function () {
            $ws = $this->wsSvc->newSpotPublicWs();
            $ws->start()->then(function () use ($ws) {
                $ws->ticker(['ETH-USDT', 'BTC-USDT'], fn() => null)->then(function ($id) {
                    sleep(self::STEP);
                    return $id;
                })->then(function ($id) use ($ws) {
                    return $ws->unsubscribe($id);
                })->then(function () use ($ws) {
                    return $ws->stop();
                })->then(function () {
                    echo "WS STAR/STOP [OK]\n";
                }, function ($err) {
                    $this->wsErrCnt++;
                    echo "WS STAR/STOP [ERROR] {$err->getMessage()}\n";
                });
            });
        });
    }

    private function statLoop(): void
    {
        $this->loop->addPeriodicTimer(self::STEP, function () {
            printf("Stat Market_ERR:%d WS_SS_ERR:%d WS_MSG:%d\n",
                $this->mkErrCnt, $this->wsErrCnt, $this->wsMsgCnt);
        });
    }
}

(new Runner())->run();

File: sdk/java/example/src/main/java/com/kucoin/test/regression/RunForeverTest.java
package com.kucoin.test.regression;

import com.kucoin.universal.sdk.api.DefaultKucoinClient;
import com.kucoin.universal.sdk.api.KucoinClient;
import com.kucoin.universal.sdk.api.KucoinWSService;
import com.kucoin.universal.sdk.generate.spot.market.GetAllSymbolsReq;
import com.kucoin.universal.sdk.generate.spot.market.GetAllSymbolsResp;
import com.kucoin.universal.sdk.generate.spot.market.MarketApi;
import com.kucoin.universal.sdk.generate.spot.spotpublic.SpotPublicWs;
import com.kucoin.universal.sdk.model.ClientOption;
import com.kucoin.universal.sdk.model.Constants;
import com.kucoin.universal.sdk.model.TransportOption;
import com.kucoin.universal.sdk.model.WebSocketClientOption;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import lombok.extern.slf4j.Slf4j;

/**
 * Runner class to periodically fetch market data, manage WebSocket subscriptions, and track
 * statistics.
 */
@Slf4j
public class RunForeverTest {

  private static final double STEP = 5.0; // seconds

  private AtomicInteger wsMsgCnt = new AtomicInteger(0);
  private AtomicInteger wsErrCnt = new AtomicInteger(0);
  private AtomicInteger mkErrCnt = new AtomicInteger(0);

  private final KucoinWSService wsSvc;
  private final MarketApi marketApi;
  private final ScheduledExecutorService executor;

  public RunForeverTest() {
    this.executor = Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors());
    ClientOption clientOption =
        ClientOption.builder()
            .key(System.getenv("API_KEY"))
            .secret(System.getenv("API_SECRET"))
            .passphrase(System.getenv("API_PASSPHRASE"))
            .transportOption(TransportOption.builder().keepAlive(true).build())
            .websocketClientOption(WebSocketClientOption.defaults())
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .build();

    KucoinClient client = new DefaultKucoinClient(clientOption);
    this.wsSvc = client.getWsService();
    this.marketApi = client.getRestService().getSpotService().getMarketApi();
  }

  public void run() {
    try {
      marketLoop();
      wsForever();
      wsStartStopLoop();
      statLoop();

      // Keep the application running
      CountDownLatch latch = new CountDownLatch(1);
      latch.await();
    } catch (InterruptedException e) {
      log.error("Main thread interrupted", e);
      Thread.currentThread().interrupt();
    } finally {
      executor.shutdown();
    }
  }

  private void marketLoop() {
    executor.scheduleAtFixedRate(
        () -> {
          try {
            GetAllSymbolsResp resp =
                marketApi.getAllSymbols(GetAllSymbolsReq.builder().market("USDS").build());
            log.info("MARKET API [OK] {}", resp.getData().size());
          } catch (Exception e) {
            mkErrCnt.incrementAndGet();
            log.info("MARKET API [ERROR] " + e.getMessage());
          }
        },
        0,
        (long) STEP,
        TimeUnit.SECONDS);
  }

  private void wsForever() {
    SpotPublicWs ws = wsSvc.newSpotPublicWS();
    ws.start();
    ws.orderbookLevel50(
        new String[] {"ETH-USDT", "BTC-USDT"}, (____, __, ___) -> wsMsgCnt.incrementAndGet());
  }

  private void wsStartStopLoop() {
    executor.scheduleAtFixedRate(
        () -> {
          try {

            SpotPublicWs ws = wsSvc.newSpotPublicWS();
            ws.start();
            String id = ws.ticker(new String[] {"ETH-USDT", "BTC-USDT"}, (__, ___, ____) -> {});
            Thread.sleep(1000);
            ws.unSubscribe(id);
            ws.stop();
            log.info("WS START/STOP [OK]");
          } catch (Exception e) {
            wsErrCnt.incrementAndGet();
            log.info("WS START/STOP [ERROR] " + e.getMessage());
          }
        },
        0,
        (long) STEP,
        TimeUnit.SECONDS);
  }

  private void statLoop() {
    executor.scheduleAtFixedRate(
        () -> {
          log.info(
              "Stat Market_ERR:{} WS_SS_ERR:{} WS_MSG:{}",
              mkErrCnt.get(),
              wsErrCnt.get(),
              wsMsgCnt.get());
        },
        0,
        (long) STEP,
        TimeUnit.SECONDS);
  }

  public static void main(String[] args) {
    new RunForeverTest().run();
  }
}
